all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+  * `TEMPORAL_CONFIG_FILE` - the path to the config file. The default is `~/.config/temporalio/temporal.json`.
+* ❓ How to do grpc-meta as env var? Like for `My-Header: My-Value`
+  * Env var names are case-insensitive and don't use dashes, so we can't put header name in env var name. And we may not
+    want to canonicalize the header names even though most gRPC clients do it for you.
+  * There's no obvious env var approach to array of strings, and it can't be completely comma-delimited because header
+    values often contain commas. Would rather not force users to have an index in the env var name. Should we just not
+    support gRPC meta from env var until we think this through?
+
+## Implementation
+
+### CLI
+
+* CLI will accept `--profile`.
+  * This deprecates `--env`/`--env-file`
+    * 💭 Why deprecate?
+      * The `env` name is confusing with environment variables. Note we do not use the term "environment" anywhere in
+        this document except for environment variables.
+      * The `env` setting uses YAML files in `~/.config/temporalio/temporal.yaml`.
+    * Error if both options are present, or in the case of default, if there are default files for both.
+      * 💭 Why error in the case of both default env and profile being present instead of merge?
+        * This is the new way and we should not try to support both at the same time, it can get confusing.
+  * The default is `default`, and this can also be set via `TEMPORAL_PROFILE` env var.
+* CLI will accept `--config-file`.
+  * Default is `~/.config/temporalio/temporal.json`, also can be set via `TEMPORAL_CONFIG_FILE` env var.
+* CLI will have a whole new set of `config` commands that operate similar to `env` commands today.
+  * Will not go into detail in this proposal, but it's very similar.
+  * We will strictly validate keys when setting whereas we did not with `env`.
+* Unlike `env`/`--env`, this does not blindly apply any field as any CLI argument.
+  * 💭 Why?
+    * We want strict validation of config values. The idea that you can set a `workflow-id` as part of this profile does
+      not make sense. These are client options only, declarative options for other things can be discussed separately.
+* CLI should defer to the Go SDK for loading/using the config/profile.
+
+### SDKs
+
+SDKs must support loading client configuration from optional profile name and optional config name. This needs to
+include a way to make sure that user-defaults can be set.
+
+Possible language-specific ideas are below. Note, they are not exact and the naming is all subject to change, they are
+simply ideas for discussion. Many of the approaches below are based on several rewritings after seeing how they don't
+work in some SDKs. The overall approach is to just make client and connection options loadable from configuration
+without overthinking shortcuts or merging or hierarchies.
+
+#### Go SDK Idea
```

Comment:
This is unnecessarily expanding the scope at first IMO...

> newConfigWithDefaults()

That's just creating a new Go config

> mergeConfig()

We have explicitly avoided any merging logic exposed to users because we do not support merging of configs because earlier versions of this proposal tried and it became too complicated

> readFromEnv()

This is what loading while disabling file does

> readFromFile()

This is what loading while disabling env does. And regardless, this is why we chose proto JSON. So people can use proto reading, writing, merging, etc.

> "build the config yourself"

Yes, this is one of the motivators behind proto JSON. You have a typesafe representation of the config you can build yourself or do anything you want with.

Reply:
I don't agree this is expanding scope—you have just said we already have 4/5 of the things I'm asking for but presented in a different (IMO less composable) form. And the 5th, `mergeConfig()`, is something we will have to implement internally anyway if we want consistent behavior when an env variable overrides a file, or when a file overrides built-in defaults. It just needs to be exposed.

Reply:
> you have just said we already have 4/5 of the things I'm asking for

I was afraid of the "etc" there and other potential things that may be requested before an initial version is delivered

> And the 5th, mergeConfig(), is something we will have to implement internally anyway if we want consistent behavior when an env variable overrides a file, or when a file overrides built-in defaults. It just needs to be exposed.

The initial version did have advanced merging logic where you could merge two configs, but this was hard to reason about. Specific variables override specific variables, this is not the same as merging though. With configuration options, especially in programmatic form, there is not a difference between unset and false. So how do you know if someone turned off TLS? Can you write an example implementation of `mergeConfig` using Go structures or proto objects? You may find it is challenging because merging programmatic objects is not the same as overwriting specifically set values that are specifically provided. The difference is subtle but important IMO. This is why we require that you create a new config from the configuration instead of "apply" configuration to an existing object.